caliban_toolbox/dataset_builder.py

```diff
@@ -0,0 +1,571 @@
+                y_new = label(y[i, :, :, 0])
+            else:
+                y_new, _, _ = relabel_sequential(y[i, :, :, 0])
+
+            y_new = remove_small_objects(y_new, min_size=small_object_threshold)
+
+            unique_objects = len(np.unique(y_new)) - 1
+            if unique_objects < min_objects:
+                keep_idx[i] = False
+
+            cleaned_y[i, :, :, 0] = y_new
+
+        # subset all dict members to include only relevant images
+        cleaned_y = cleaned_y[keep_idx]
+        cleaned_X = X[keep_idx]
+        cleaned_tissue = tissue_list[keep_idx]
+        cleaned_platform = platform_list[keep_idx]
+
+        cleaned_dict = {'X': cleaned_X, 'y': cleaned_y, 'tissue_list': list(cleaned_tissue),
+                        'platform_list': list(cleaned_platform)}
+
+        return cleaned_dict
+
+    def build_dataset(self, tissues='all', platforms='all', output_shape=(512, 512), resize=False,
+                      data_split=(0.8, 0.1, 0.1), seed=0, **kwargs):
+        """Construct a dataset for model training and evaluation
+
+        Args:
+            tissues: which tissues to include. Must be either a list of tissue types,
+                a single tissue type, or 'all'
+            platforms: which platforms to include. Must be either a list of platform types,
+                a single platform type, or 'all'
+            output_shape: output shape for dataset. Either a single tuple, in which case
+                train/va/test will all have same size, or a list of three tuples
+            resize: flag to control resizing the input data.
+                Valid arguments:
+                    - False. No resizing
+                    - by_tissue. Resizes by median cell size within each tissue type
+                    - by_image. REsizes by median cell size within each image
+            data_split: tuple specifying the fraction of the dataset for train/val/test
+            seed: seed for reproducible splitting of dataset
+            **kwargs: other arguments to be passed to helper functions
+
+        Returns:
+            list of dicts containing the split dataset
+
+        Raises:
+            ValueError: If invalid tissues argument specified
+            ValueError: If invalid platforms argument specified
+        """
+        if self.all_tissues == []:
+            self._identify_tissue_and_platform_types()
+
+        # validate inputs
+        if isinstance(tissues, list):
+            for tissue in tissues:
+                if tissue not in self.all_tissues:
+                    raise ValueError('{} is not one of {}'.format(tissue, self.all_tissues))
+        elif tissues == 'all':
+            tissues = self.all_tissues
+        elif tissues in self.all_tissues:
+            tissues = [tissues]
+        else:
+            raise ValueError(
+                'tissues should be "all", one of {}, or a list '
+                'of acceptable tissue types'.format(self.all_tissues))
```

Comment:
These 2 bits of logic look so similar, it could probably be a `validate` function that works for both tissues and platforms.